pkg/trace/writer/trace.go

```diff
@@ -148,65 +156,49 @@ func (w *TraceWriter) Stop() {
 	w.sender.Stop()
 }
 
-func (w *TraceWriter) handleSampledTrace(sampledTrace *TracePackage) {
-	if sampledTrace == nil || sampledTrace.Empty() {
+func (w *TraceWriter) handleSampledTrace(pkg *TracePackage) {
+	if pkg == nil || pkg.Empty() {
 		log.Debug("Ignoring empty sampled trace")
 		return
 	}
 
-	trace := sampledTrace.Trace
-	events := sampledTrace.Events
+	apiTrace, size := traceutil.APITrace(pkg.Trace)
+	size += pb.Trace(pkg.Events).Msgsize()
 
-	n := len(trace) + len(events)
-
-	if w.spansInBuffer > 0 && w.spansInBuffer+n > w.conf.MaxSpansPerPayload {
-		// If we have data pending and adding the new data would overflow max spans per payload, force a flush
-		w.flushDueToMaxSpansPerPayload()
+	if w.payloadSize > 0 && w.payloadSize+size > payloadFlushThreshold {
+		// this new package would push us over, flush
+		log.Debug("Flushing... Reached size threshold.")
+		w.flush()
 	}
 
-	w.appendTrace(sampledTrace.Trace)
-	w.appendEvents(sampledTrace.Events)
-
-	if n > w.conf.MaxSpansPerPayload {
-		// If what we just added already goes over the limit, report this but lets carry on and flush
-		atomic.AddInt64(&w.stats.SingleMaxSpans, 1)
-		w.flushDueToMaxSpansPerPayload()
+	log.Tracef("Handling new trace with %d spans: %v", len(pkg.Trace), pkg.Trace)
+	w.traces = append(w.traces, apiTrace)
+	if len(pkg.Events) > 0 {
+		log.Tracef("Handling new APM events: %v", pkg.Events)
+		w.events = append(w.events, pkg.Events...)
 	}
-}
-
-func (w *TraceWriter) appendTrace(trace pb.Trace) {
-	numSpans := len(trace)
+	w.payloadSize += size
+	w.spansInBuffer += len(pkg.Trace) + len(pkg.Events)
 
-	if numSpans == 0 {
-		return
+	if size > payloadMaxSize {
+		log.Warnf("Trace might be dropped by API; size exceeds maximum (%d): %d", payloadMaxSize, size)
 	}
-
-	log.Tracef("Handling new trace with %d spans: %v", numSpans, trace)
-
-	w.traces = append(w.traces, traceutil.APITrace(trace))
-	w.spansInBuffer += numSpans
-}
-
-func (w *TraceWriter) appendEvents(events []*pb.Span) {
-	for _, event := range events {
-		log.Tracef("Handling new APM event: %v", event)
-		w.events = append(w.events, event)
+	if size > payloadFlushThreshold {
+		// we've added a single package that surpasses our threshold, we should count this occurrence,
+		// it could be an indication of "over instrumentation" on the client side, where too many spans
+		// could be present in traces.
+		atomic.AddInt64(&w.stats.SingleMaxSize, 1)
+		log.Debugf("Flushing... Surpassed size threshold with a single package sized %d bytes", size)
+		w.flush()
```

Comment:
I guess you could be right, but I'd say that's outside the scope of this PR because the original logic was like this. If we would remove that, I'd like to add better more thorough testing to ensure nothing breaks. Plus, why not flush it immediately instead of waiting for the next run?

Reply:
> Plus, why not flush it immediately instead of waiting for the next run?

Only reason is for simplicity, so there's just one place where we trigger a flush. Not a big deal though, this is also fine.

Reply:
I'm not against it at all. Not sure if @AlexJF had any reasons for this form.

Reply:
Has been a while but I don't think there was a strong reason for this other than getting data out of the way ASAP (i.e., instead of waiting until a new payload comes in or the flush timeout to be reached, flush immediately since we know this "chunk" is done).

Reply:
It's an edge case. If this were the first iteration of this writer, I might've voted against it, but now we have it and I don't see why we'd remove it. It's clear that this condition is met sometimes based on seeing the metric on customer accounts. Let's keep it.

Reply:
👍